### ISSUE TITLE
Dexcom Clarity Uploader

### ISSUE THREAD

**Issue by PatrikTrestik** [NightscoutFoundation/xDrip#229]
When using upload to Dexcom servers it works with Dexcom Follow mobile app, but no data are visible in Clarity web app.

**Comment by KaledDiabetic** [comment]
Hi, I'm facing the same issue, someone said there is 3 hours delay. I have been waiting more than 10 hours and still no data in Dexcom Clarity, in the mean time Dexcom Follow works fine. Have you fixed the problem?
Thanks

**Comment by PatrikTrestik** [comment]
My problem still persists. 

Data definitely reached some dexcom server because it was distributed to Follow mobile app.

I would like to sniff http going out of official Dexcom App, but I haveit  only on iPhone and it is not easy there.

**Comment by musicman0329** [comment]
xdrip+ _does not_ send data to Dexcom Clarity. As far as I know, there are no plans to include this functionality into xdrip+ for a variety of reasons.

As far as the three hours delay, that relates to the iPhone G5 Mobile App sending data to Apple Health, and has nothing to do with Clarity.

**Comment by PatrikTrestik** [comment]
For me it is not problem as I use Nightwatch for storing and viewing my son's data.

But it is problem for our clinic. They are fixed to Diasend for all patients data.
Only way I know how to upload there is Clarity -> Diasend connect.
If someone know about other, I'm happy to investigate.

**Comment by musicman0329** [comment]
There is no way to upload data from xdrip+ or Nightwatch into Clarity or Diasend. You could use Nightscout instead, which has it's own full-featured reporting tool set.

**Comment by PatrikTrestik** [comment]
As I said Nightwatch is great for me, but clinic requires Diasend.
It looks like it will need some coding. I have some spare time over holiday and will lookup my options.
I'll share with community if there is some outcome.

**Comment by wicky1988** [comment]
Hi all,

Also desperately looking for this feature, wondering if you've made any progress @PatrikTrestik? Happy Holidays!

**Comment by PatrikTrestik** [comment]
I have working prove of concept for Diasend uploader.
Same work can be done with Clarity too.
It'll definitely require communicating with target companny (my target is 
Diasend). If not for teechnical, surely for licence reasons.

I already communicate with Diasend.

Is your prime interest Clarity or other cloud service?

Patrik


Dne 31. prosince 2017 9:34:59 PM wicky1988 <notifications@github.com> napsal:

**Comment by wicky1988** [comment]
Prime interest is definitely Clarity.

**Comment by SuicidalTissue** [comment]
+1 also req. this - I'm having to carry around and calibrate both xdrip and a g5 reciever so I can eventually get the stats into diasend for clinic to use

**Comment by PatrikTrestik** [comment]
Just report on progress.

Diasend - waiting for response from them regarding protocol doc. Upload is binary data so no chance to do it without their help.

Clarity - no answer from Dexcom. But protocol is more open. I see chance here. I need more time :(

**Comment by beteez** [comment]
@PatrikTrestik  any chance i could try to help you out on this? id love to start contributing, but i dont want to reinvent the wheel if you've done all the legwork.

**Comment by SuicidalTissue** [comment]
Hey @beteez @PatrikTrestik, sorry to be naggy, but any update to this feature request?

**Comment by PatrikTrestik** [comment]
Sorry, I had the overload of my job responsibilities. All my private projects ware postponed for a whole month.
I hope I'm back on it this month.

**Comment by Pinx0** [comment]
I'm also interested in this feature. My clinic requires diasend but I also like clarity to review the data. Clarity goes into diasend so no issue there.

**Comment by dabear** [comment]
Not possible currently. Clarity protocol is much stricter than dexcom share. It requires much more of a reversing job, as they do and integrity protection using hmac-md5 on the uploaded data. We need to know how this integrity protection key is generated and applied before we can recreate it. @miguelkennedy has been looking into this, recently.

**Comment by miguelkennedy** [comment]
Correct... still need to connect my G5 again to the Dexcom app after a clean install and debug communications to check if the key to generate the HMAC is sent to the app when the user logins for the first time... It's already on my TODO list.

**Comment by norbertgaal** [comment]
+1, this would be really nice (even if it's just a script that uploads NS data to Clarity or something like that)

**Comment by PatrikTrestik** [comment]
Hi. My project got stuck on licensing and legal problem. Dexcom has no public API for writing data to Clarity. Obviously, there is good API but locked for their own apps.
I have done reverse engineering and I have partially working uploader. But it is licensing breach if I'll make it public.
Dexcom is not communicating with me much. I expect that they have no interest in open source.
I'll keep trying while I finish uploader to beta stage.

**Comment by Oekonom2** [comment]
Thank you for your time invested in this feature. It is very sad that Dexcom is not cooperating in this matter.

**Comment by darenfit** [comment]
I noticed that the Diasend uploader application interfaces with a number of other meters.  Picking the freestyle Neo as an example I found some work already done on the interface format https://github.com/Flameeyes/glucometer-protocols/blob/master/abbott/freestyle-precision-neo.md

Can anyone see a possible way to upload to Diasend by emulating another device?  There seems to be a possibility of  creating a virtual HID device or maybe a solution as simple as uploading to one of the accepted meters and then uploading from the meter using the official application.

My aim is purely to keep my endo informed at the review meetings, for my own personal use nightscout is preferred.

**Comment by darenfit** [comment]
I noticed that the Diasend uploader application interfaces with a number of other meters. Picking the freestyle Neo as an example I found some work already done on the interface format https://github.com/Flameeyes/glucometer-protocols/blob/master/abbott/freestyle-precision-neo.md

Can anyone see a possible way to upload to Diasend by emulating another device? There seems to be a possibility of creating a virtual HID device or maybe a solution as simple as uploading to one of the accepted meters and then uploading from the meter using the official application.

My aim is purely to keep my endo informed at the review meetings, for my own personal use Nightscout is preferred.

**Comment by bassettb** [comment]
This would be a big help for us also so I'd like to help with the development and testing.  Is there a branch out there I can look at?

**Comment by acornett** [comment]
@PatrikTrestik  - IANAL, but I'd be interested to hear why you think that your reverse engineered code would be a legal issue.  The DCMA has [reverse engineering exemptions](https://www.ftc.gov/news-events/blogs/techftc/2016/10/dmca-security-research-exemption-consumer-devices) for consumer devices, and that includes most medical devices.

As long as you've not signed an NDA with Dexcom you may be able to publish your findings.  If you're concerned about the specific code, you could always describe in more general terms your findings regarding the data format, any checksums, etc and some else could prepare a clean implementation.

**Comment by nicolajbuthler** [comment]
+1 on upload to Diasend. @PatrikTrestik Would be great to hear about your workaround to Clarity 👍
 
Diasend/Glooko receives information from different kind of glucose meters, so by uploading xDrip+ data to a glucosemeter and then to Diasend might be a solution.

Need to upload my glucosedata to Diasend, so my clinic can get the information where they as well can see data from my Omnipod.

**Comment by MickBruno** [comment]
Apologies if this is not strictly on-topic, but as a temporary workaround to pushing data to Clarity is there a way to simply download the data from xDrip to a CSV file? Looking thru the xDrip menus I couldn't find anything.

**Comment by haspden** [comment]
@patriktrestik any updates on this? I've been trying for months to come up with an outside of the box solution to get around this

**Comment by PatrikTrestik** [comment]
Hi. I'm really sorry not to move the project for a long time. It is all my personal life problem so I'll not bother you with it here.
Only relevant information could be that dust settles down and I might come back to coding in next month or so.
Thanks for patience.

**Comment by haspden** [comment]
@nicolajbuthler  I can get it into diasend using a meter, however it comes up as a million blood test results not CGM data, any use?

**Comment by nicolajbuthler** [comment]
@haspden Sounds like a solution to the problem. I dont think the endocrinologists care if it looks like CGM or blood test result in Diasend. Might as well try. What method did you use to get it uploaded?

**Comment by nicolajbuthler** [comment]
@haspden Can you give me a hint to how you did it? Meeting my endo on 30. January 😊

**Comment by Nightfoxy** [comment]
Any more movement on this?

**Comment by tbower3** [comment]
I could really use an upload to diasend feature too.  As they are open to so many meters, and it is to their benefit to make their platform easy to use, surely they would help with the parameters of upload from x-drip?

**Comment by trbarb** [comment]
Hoping the Dexcom upload issue can be resolved.  Same as another commenter, my Endo uses Dexcom Clarity apps.  I like XDrip, but have to be about to get my reading to the doc.  May be able to find a workaround, but as of now, I'm having to go back to the Dexcom app.

**Comment by dabear** [comment]
This is not a bug. It's a feature request to support uploading to a manufacturers (dexcom) web service. Dexcom explicitly allows only its own app to do this. To support this will require reverse engineering, and it's not an easy just-so-it kind of task.

**Comment by PatrikTrestik** [comment]
Hi to all. I would like to resurrect this issue/project "How to upload data to Diasend/Clarity.
If you are still listening to this thread, please share your ideas and information about the current situation.

**Comment by tbower3** [comment]
Hi, very interested in uploading from nightscout to Diasend - maybe Diasend could help, especially if they are told how many people are using nightscout around the world?

**Comment by PatrikTrestik** [comment]
I started a new project, see status updates here https://github.com/PatrikTrestik/diasend-upload/

**Comment by charli144** [comment]
Thanks for your hard work. We are all waiting for this.

**Comment by tolot27** [comment]
Please post/comment Diasend related stuff to #716. This issue requests Clarity uploads.

Because we try to improve our issue management to speed up development, it would be really helpful to have clearly focused and "clean" issues, comments included.

**Comment by Navid200** [comment]
@tolot27 Can xDrip upload to Clarity (the title)?  I don't think so.
Shouldn't we close this request because it currently is not possible due to the limitation imposed by Dexcom?
There is nothing we can do to xDrip to resolve this, is there?

**Comment by tolot27** [comment]
This is a valid feature request and should not be closed until someone made a final decision.

**Comment by tolot27** [comment]
Because no public API is available, we are closing this issue for now.

**Comment by Navid200** [comment]
https://navid200.github.io/xDrip/docs/Dexcom/Clarity.html
